doc: setup logical parition scenario on kbl nuc i7

leilu opened the pull request:
This patch adds tutorial about how to setup logical partition scenario on nuc7i7dnb

Signed-off-by: Lei Lu <leix.lu@intel.com>

wenlingz commented:
@wuxyintel , please help review it, thanks

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+
+Disable "Intel Hyper Threading Technology" in the BIOS.
+
+In the logical partition scenario, two pre-launched VMs (running Clear Linux) are started by the ACRN hypervisor. To set up the Linux root filesystems for each VM, follow the Clear Linux OS bare metal installation guide to install Clear Linux OS on a SATA disk and a USB flash disk prior the setup, as the two pre-launched VMs will mount the root file systems via the SATA controller and the USB con ...
```
@leilu please use "a storage device with USB interface  (such as USB Flash or SATA disk connecting with USB3.0 SATA converter etc)" instead of "a USB flash", and check it in this whole document.

gvancuts replied:
Could someone use a SATA and NVMe drive instead for example? I believe the underlying requirement is to use controllers that are distinct PCI devices, is that correct?

wuxyintel replied:
@gvancuts Yes, the requirement is to use SATA controller and USB controller.

coderlei replied:
fix at doc v2 version

leilu replied:
fix at doc v2 version.

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+Guidelines of Setup Logical Partition Scenario on KBL NUC i7
+############################################################
+
+ACRN hypervisor supports logic partition scenario, in which the User OS (such as Clear Linux) running in a pre-launched VM can bypass the ACRN hypervisor and directly access isolated PCI devices. The following guidelines provide step by step instructions on how to set up the ACRN hypervisor logical partition scenario on KBL NUC i7 boards running two pre-launched VMs. The logical partition scenari ...
```
Please use "NUC7i7DNB (is called KBL NUC i7) " instead of the first "KBL NUC i7". Thanks!

coderlei replied:
fix at doc v2 version

leilu replied:
fix at doc v2 version.

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+
+
+Note: If you update kernel image (bzImage) and modules on ubuntu of KBL NUC i7, need use scp to copy it from the remote host building machine.
+If you install EBtool Docker on the ubuntu of KBL NUC i7, it will be easy to update kernel image and modules.
```
Where is EBtool Docker? Can you help to make it clearer?

coderlei replied:
Remove EBtool Docker for doc v2 version, as the the EBtool docker is only internal used.

leilu replied:
Remove EBtool Docker for doc v2 version, as the the EBtool docker is only internal used.

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+
+  $ git clone https://github.com/projectacrn/acrn-hypervisor.git
+  $ cd acrn-hypervisor/hypervisor
+  $ make menuconfig
```
Please use "make menuconfig BOARD=nuc7i7bnh" instead of "make menuconfig", and double check the following configuration items. Thanks!

gvancuts replied:
If I followed you correctly, you are building the hypervisor in Ubuntu. If true, we should also refer to the "Build ACRN from source" (using :ref:`getting-started-building`) for further instructions on how to correctly set-up Ubuntu to build ACRN (as there are a few tricks needed if using 18.04 or 16.04)

wuxyintel replied:
@gvancuts Thanks, let's follow "Build ACRN from source" and refer it. use "make defconfig BOARD=nuc7i7bnh" and "make menuconfig" here.

wuxyintel replied:
@gvancuts thanks! let's follow "Build ACRN from source" and refer it; use "make defconfig BOARD=nuc7i7bnh" and "make menuconfig"

coderlei replied:
fix at doc v2 version

leilu replied:
fix at doc v2 version

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+  $make BOARD=nuc7i7bnh
+
+Note: If you update hypervisor image (``acrn.32.out``) on ubuntu of KBL NUC i7, need use scp to copy it from the remote host building machine.
+If you install EBtool Docker on the ubuntu of KBL NUC i7, it will be easy to update hypervisor image.
```
Where is EBtool Docker and how to install it?

leilu replied:
Remove EBtool Docker for doc v2 version, as the the EBtool docker is only internal used.

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+
+Step 1: Append the following configuration to the ``/etc/grub.d/40_custom`` file::
+
+  menuentry 'ACRN hypervisor Logical Partition Scenario' --class ubuntu --class gnu-linux --class gnu --class os $menuentry_id_option 'gnulinux-simple-e23c76ae-b06d-4a6e-ad42-46b8eedfd7d3' {
```
Please add note to double check the title of this new grub boot entry should be the same the other normal entry.

leilu replied:
add at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+.. _setup_logical_partition_scenario_on_kbl_nuc_i7:
+
+Guidelines of Setup Logical Partition Scenario on KBL NUC i7
```
Any reason this would only work on Core i7 processors?

If not, let's be more generic throughout the doc and simply indicate that these steps were validated on the NUC you used.

wuxyintel replied:
@gvancuts for the current vm configuration of logical partition, the physical board should have 4 physical cores, core i7 is selected.

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+Guidelines of Setup Logical Partition Scenario on KBL NUC i7
+############################################################
+
+ACRN hypervisor supports logic partition scenario, in which the User OS (such as Clear Linux) running in a pre-launched VM can bypass the ACRN hypervisor and directly access isolated PCI devices. The following guidelines provide step by step instructions on how to set up the ACRN hypervisor logical partition scenario on KBL NUC i7 boards running two pre-launched VMs. The logical partition scenari ...
```
Can you break this down in lines no longer than 80 characters?

@dbkinder will do a more thorough review of the text but I'm not sure the expression "by-pass the hypervisor" is the best. How about we rephrase this into something like "has direct access to specific PCI devices".?

leilu replied:
fix at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+Guidelines of Setup Logical Partition Scenario on KBL NUC i7
+############################################################
+
+ACRN hypervisor supports logic partition scenario, in which the User OS (such as Clear Linux) running in a pre-launched VM can bypass the ACRN hypervisor and directly access isolated PCI devices. The following guidelines provide step by step instructions on how to set up the ACRN hypervisor logical partition scenario on KBL NUC i7 boards running two pre-launched VMs. The logical partition scenari ...
```
There is a mix of `logic partition` and `logical partition` in the text, can you use `logical` throughout the doc?

leilu replied:
fix at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+Prerequisites
+*************
+
+Disable "Intel Hyper Threading Technology" in the BIOS.
```
Can we add a short explanation as to why we make this recommendation (or is strictly required)?

wuxyintel replied:
@gvancuts There are two reasons: one is to mitigate L1TF, other is to avoid interference on the logical partition scenario between logical cores.

leilu replied:
add doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+Update kernel image and modules of Pre-launched VM
+**************************************************
+
+Step1: On your development workstation, clone the ACRN kernel source tree, and build the Linux kernel image that will be used to boot the privileged VMs::
```
Can you use a numbered list here? (see https://projectacrn.github.io/latest/developer-guides/doc_guidelines.html#lists)

leilu replied:
fix at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+  $ sudo cp <path-to-kernel-image-built-in-step1>/bzImage /boot/
+
+
+Note: If you update kernel image (bzImage) and modules on ubuntu of KBL NUC i7, need use scp to copy it from the remote host building machine.
```
Use `.. note::`

leilu replied:
fix at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+    $dmesg | grep ttyS0
+    [    0.000000] console [ttyS0] enabled
+    [    1.562546] 00:01: ttyS0 at I/O 0x3f8 (irq = 4, base_baud = 115200) is a 16550A
+    The following command prints detailed information about all PCI buses and devices in the system.
```
Break out of the command block

leilu replied:
fix at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+    [    0.000000] console [ttyS0] enabled
+    [    1.562546] 00:01: ttyS0 at I/O 0x3f8 (irq = 4, base_baud = 115200) is a 16550A
+    The following command prints detailed information about all PCI buses and devices in the system.
+    $lspci -vv
```
```suggestion
    $ lspci -vv
```

Do you need `sudo`?

wuxyintel replied:
@gvancuts  Yes, we need "sudo", thanks!

leilu replied:
fix at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+            Subsystem: Intel Corporation Sunrise Point-LP SATA Controller [AHCI mode]
+    00:1f.6 Ethernet controller: Intel Corporation Ethernet Connection I219-LM (rev 21)
+            Subsystem: Intel Corporation Ethernet Connection I219-LM
+    Note: Double check PCI devices BDF defined in the "hypervisor/arch/x86/configs/nuc7i7bnh/pci_devices.h".
```
Use `.. note::`

leilu replied:
fix at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+            Subsystem: Intel Corporation Sunrise Point-LP SATA Controller [AHCI mode]
+    00:1f.6 Ethernet controller: Intel Corporation Ethernet Connection I219-LM (rev 21)
+            Subsystem: Intel Corporation Ethernet Connection I219-LM
+    Note: Double check PCI devices BDF defined in the "hypervisor/arch/x86/configs/nuc7i7bnh/pci_devices.h".
```
```suggestion
    Note: Double check PCI devices BDF defined in the ``hypervisor/arch/x86/configs/nuc7i7bnh/pci_devices.h``.
```

leilu replied:
fix at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+
+Step 5: Build the ACRN hypervisor and copy the artifact ``acrn.32.out`` to the ``/boot`` directory::
+
+  $make BOARD=nuc7i7bnh
```
```suggestion
  $ make BOARD=nuc7i7bnh
```

leilu replied:
fix at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+
+  $make BOARD=nuc7i7bnh
+
+Note: If you update hypervisor image (``acrn.32.out``) on ubuntu of KBL NUC i7, need use scp to copy it from the remote host building machine.
```
Use `.. note::`

gvancuts replied:
I don't quite understand this note, do you mean that someone needs to transfer this `acrn.32.out` file from their dev machine in case it wasn't built in Ubuntu installed on the NVMe drive?

wuxyintel replied:
@gvancuts yes, someone builds hypervisor on remote dev machine, this file should be copied on the target board.

leilu replied:
fix at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+  update-grub -v
+  grub-mkconfig (GRUB) 2.02~beta2-36ubuntu3.18
+
+In this case, the GRUB loader name shall have ``GRUB`` prefix;
```
I'm not too sure what to do in this step, there is no Grub info in that file at the moment.

wuxyintel replied:
@gvancuts We need to double check the GRUB boot loader name which should have "GRUB" prefix, otherwise, ACRN hypervisor will fail to boot.

gvancuts replied:
Where do you check that? Can you give a more explicit example?

wuxyintel replied:
@gvancuts ACRN hypervisor checks this on the function "init_vboot_operations", please refer to https://github.com/projectacrn/acrn-hypervisor/blob/master/hypervisor/boot/guest/vboot_wrapper.c

gvancuts replied:
Sorry but I'm still completely confused about this step. You ask people to check the board specific configuration in `hypervisor/arch/x86/configs/nuc7i7bnh/board.c`, specifically for the "GRUB loader name" but that file (see [here](https://github.com/projectacrn/acrn-hypervisor/blob/master/hypervisor/arch/x86/configs/nuc7i7bnh/board.c)) has no such parameter. So what are people suppose to check??

wuxyintel replied:
Thanks for your finding. There are some mistakes in this step. It should be "Step 3: Check GRUB boot loader name." we only need to check the GRUB boot loader name on the ubuntu of target board.

leilu replied:
fix at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+   The logical partition scenario on KBL NUC i7
+
+Prerequisites
+*************
```
One of the prerequisite is also to have the following HW available:
* NVMe disk
* SATA disk
* USB stick

leilu replied:
added at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+
+Step 1: Check whether ACRN hypervisor startup fails or not, this is similar to SDC scenario on KBL NUC.
+
+Step2: Check whether two pre-launched VMs startup fails or not (such as enter user OS shell command line, devices operations testing).
```
I don't understand this line, can you be more explicit?

leilu replied:
updated at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+Logical Partition Scenario Startup Checking
+*******************************************
+
+Step 1: Check whether ACRN hypervisor startup fails or not, this is similar to SDC scenario on KBL NUC.
```
Are there three OSs running at this stage then? What OS/VM are you seeing by default (if everything runs correctly)?

leilu replied:
updated at doc v2 version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_kbl_nuc_i7.rst:
```diff
@@ -0,0 +1,154 @@
+Step 1: Check whether ACRN hypervisor startup fails or not, this is similar to SDC scenario on KBL NUC.
+
+Step2: Check whether two pre-launched VMs startup fails or not (such as enter user OS shell command line, devices operations testing).
+Note: Enter pre_launched VM from hypervisor through the hypervisor shell command ``vm_console vm_id`` and find vm_id through hypervisor shell command ``vm_list``.
```
Can you add a pointer to the ACRN Shell user guide (using :ref:`acrnshell`)

leilu replied:
added at doc v2 version

acrnsi commented:
Ok to verify

acrnsi commented:
Ok to verify

HuiHuangShi commented:
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

acrnsi commented:
Ok to verify

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,264 @@
+     *  Set Base address of serial PIO region as "0x3f8";
+     *  Set 2M-aligned Start physical address of the RAM region used by the
+        hypervisor as "0x6a000000";
+     *  Select "Enable ACPI DMA Remapping tables parsing".
```
@leilu These hypervisor configuration item is out of date, can you update this? Thanks

leilu replied:
fix at doc v3 version.

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,264 @@
+#. Check GRUB boot loader name. 
+
+   Double check GRUB loader name whether have "GRURB" prefix on ubuntu of target
+   board. If no, should update it:
```
@leilu Can you remove this duplicate statement? "Check GRUB boot loader name" is enough. Thanks!

leilu replied:
fix at doc v3 version.

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,264 @@
+
+   .. code-block:: none
+   
+      $ make BOARD=nuc7i7bnh
```
@leilu "$make" is formal one, can you check it and update this? Thanks!

leilu replied:
fix at doc v3 version.

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,264 @@
+
+    * To mitigate L1TF.
+    * To avoid interference on the logical partition scenario between logical
+      cores.
```
@leilu This explanation may be note, can you update it? thanks!

leilu replied:
fix at doc v3 version.

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+
+#. Clone the ACRN source code and configure the build options
+
+   Please refer :ref:`getting-started-building` to setup ACRN build envniroment
```
```suggestion
   Please refer :ref:`getting-started-building` to setup ACRN build environment
```

leilu replied:
fixed at doc v4 version.

wuxyintel replied:
Reviewed-by: Xiangyang.wu@intel.com, BTW, can you add commit message about this PR and wait @gvancuts and @dbkinder 's confirm, thanks!

wuxyintel commented:
Reviewed-by: Xiangyang.wu@intel.com, BTW, can you add commit message about this PR and wait @gvancuts and @dbkinder 's confirm, thanks!

leilu commented:
@xiaoguangwu I have added commit message about this PR.

> Reviewed-by: [Xiangyang.wu@intel.com](mailto:Xiangyang.wu@intel.com), BTW, can you add commit message about this PR and wait @gvancuts and @dbkinder 's confirm, thanks!

I have added commit message about this PR.

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+Update kernel image and modules of Pre-launched VM
+**************************************************
+#. On your development workstation, clone the ACRN kernel source tree, and build
+   the Linux kernel image that will be used to boot the privileged VMs:
```
"privliged" is a new term here.  Do we need to define it? You've been calling these "pre-launched VMs"

wuxyintel replied:
Yes, thanks, should update it as "pre-launched VMs".

leilu replied:
will fix at next version.

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+
+      $ sudo cp <path-to-kernel-image-built-in-step1>/bzImage /boot/
+
+   .. note::
```
These two notes can be combined and simplified.  At the beginning of this section you say to use a "development workstation" but here you're implying you can use your NUC as the development workstation?

```
.. note::
   Use ``cp`` as shown, if you're development workstation is the NUC, or ``scp`` if the image
   was built on another computer.
```

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+   address of the serial port, and the PCI BDF addresses of the SATA controller
+   and the USB controllers on NUC7i7DNB board. Enter the following command to
+   get the I/O addresses of the serial port. NUC7i7DNB boards support one
+   serial port ttyS0. Need to connect the serial port to the development host,
```
```suggestion
   serial port ttyS0. You'll need to connect the serial port to the development host,
```

leilu replied:
will fix at next version.

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+
+   .. note:: 
+
+      Double check PCI devices BDF defined in the ``hypervisor/arch/x86/configs/nuc7i7bnh/pci_devices.h``.
```
What should you check in this file?  How would you know if something is wrong?

wuxyintel replied:
Use comment "sudo lspci -vv" when board runs unbuntu to read PCI devices BDF, these PCI devices BDF is defined on our configuration file "pci_devices.h", check it one by one.

wuxyintel replied:
@dbkinder Do you have any more comment about this? Or how to make this clearer, thanks!

dbkinder replied:
I guess I'm asking what should you look at? Are you comparing the output of the lspci -vv command with information in the pci_devices.h file?  If so, then you could update this note by saying that:
```
Double check PCI devices BDF defined in the ``hypervisor/arch/x86/configs/nuc7i7bnh/pci_devices.h``
with the information reported by the ``lspci -vv`` command.
```

wuxyintel replied:
Thanks you to make this clearer, we will update this according to your comments.

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+     *  Set Maximum number of PCPU as "4";
+     *  Clear/Disable “Enable hypervisor relocation”.
+
+   It is recommended to keep the default values of the other items.
```
```suggestion
   We recommend keeping the default values of items not mentioned above.
```

leilu replied:
will fix at next version.

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+      grub-mkconfig (GRUB) 2.02~beta2-36ubuntu3.18
+
+   In this case, the GRUB loader name shall have ``GRUB`` prefix;
+#. Check or update BDF information of PCI devices of each pre-launched VM;
```
What needs updating?

wuxyintel replied:
PCI devices is checked in the previous step, If default pci devices BDF is configured is wrong, user needs to update it according to the results of the previous pci devices BDF checking.

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+   
+      $ make
+
+   .. note:: 
```
Merge and simplify these two notes, as shown previously.

leilu replied:
will fix at next version.

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+
+#. Reboot the NUC7i7DNB board. Select the "ACRN hypervisor Logical Partition
+   Scenario" entry to boot the logical partition of the ACRN hypervisor on the
+   NUC7i7DNB displayer, GRUB loader will boot the hypervisor, and the hypervisor
```
```suggestion
   NUC7i7DNB display. The GRUB loader will boot the hypervisor, and the hypervisor
```

leilu replied:
will fix at next version.

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+Logical Partition Scenario Startup Checking
+*******************************************
+
+#. Check whether ACRN hypervisor startup fails or not:
```
```suggestion
#. Use these steps to verify the hypervisor is properly running:
```

leilu replied:
will fix at next version.

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+
+   #. Login ACRN hypervisor shell from serial console.
+   #. Use the ``vm_list`` check pre-launched VMs.
+#. Check whether two pre-launched VMs startup fails or not:
```
```suggestion
#. Use these steps to verify the two pre-launched VMs are running properly:
```

leilu replied:
will fix at next version.

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+   #. Use the ``vm_console 1`` to switch to VM1's console.
+   #. The VM1's clearlinux OS could boot up and login in.
+
+.. note:: 
```
I'd delete this note and replace it with a sentence:
```
Refer to the :ref:`ACRN hypervisor shell user guide <acrnshell>`
for more information about available commands.
```

leilu replied:
will fix at next version.

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+.. _setup_logical_partition_scenario_on_nuc7i7dnb:
+
+Guidelines of Setup Logical Partition Scenario on NUC7i7DNB
```
@dbkinder Sorry about this name confuses you.  "NUC7i7DNHE" is listed on the https://projectacrn.github.io/latest/hardware.html#intel-kaby-lake-nuc. As we known, the intel KBL NUC has product name and board name. The platform is used here, "NUC7i7DNHE" is its product name, "NUC7i7DNB" is its board name. BTW, in fact, the board "NUC7i7DNB" is supported, there are three products for this board, please refer to page #iii of https://www.intel.com/content/dam/support/us/en/documents/mini-pcs/nuc-kits/NUC7i7DN_TechProdSpec.pdf. the board name may be precise for developer and customer. Do you have some suggestion about which name is needed to be used? Thanks!

wuxyintel replied:
The board name of  "Intel NUC Kit NUC7i7DNHE" on the hw page is “NUC7i7DNB”, can we use  "Intel NUC Kit NUC7i7DNHE" instead of NUC7i7DNB? or do you have another suggestion about this, thanks!

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+
+  .. note::
+
+     To mitigate L1TF.
```
@dbkinder yes, thanks! let's update this.

leilu replied:
will fix at next version.

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+as Clear Linux) running in a pre-launched VM can bypass the ACRN hypervisor and
+directly access isolated PCI devices. The following guidelines provide step by
+step instructions on how to set up the ACRN hypervisor logical partition
+scenario on NUC7i7DNB boards running two pre-launched VMs. The logical partition
```
NUC7i7DNB is board name, we enable logical partition is for this board, not for all NUC.

wuxyintel replied:
Yes, for current design, logical partition is only supported on "Intel NUC Kit NUC7i7DNHE" on the hw page.

leilu commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+* NVMe disk
+* SATA disk
+* Storage device with USB interface (such as USB Flash
+  or SATA disk connecting with USB3.0 SATA converter etc)
```
will fix at next version.

leilu commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+      $ make 
+      $ make modules_install INSTALL_MOD_PATH=out/
+
+   The last two commands built the bootable kernel image in the
```
will fix at next version.

leilu commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+   board later.
+
+#. Current ACRN logical partition scenario implementation requires a multi-boot
+   capable bootloader to boot both ACRN hypervisor and the bootable kernel image
```
will fix at next version.

leilu commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+     To avoid interference on the logical partition scenario between logical
+     cores.
+
+* In the logical partition scenario, two pre-launched VMs (running Clear Linux)
```
will fix at next version.

wuxyintel commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,267 @@
+
+#. Current ACRN logical partition scenario implementation requires a multi-boot
+   capable bootloader to boot both ACRN hypervisor and the bootable kernel image
+   built from the previous step. You could install Ubuntu OS to the NUC7i7DNB
```
@dbkinder yes, thanks. We install ubuntu to using GRUB bootloader. BTW, it is hard to install grub bootloader on the clear linux to load images. In the current phase, we use ubuntu to get GRUB boot loader, there may be some methods to install GRUB on the other OS or on the storage directly. We will update this according to your comment.

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,247 @@
+Prerequisites
+*************
+
+* `NUC7i7DNB <https://ark.intel.com/content/www/us/en/ark/products/130394/intel-nuc-board-nuc7i7dnbe.html>`_
```
Let's list the NUC Kit models we support here instead of the board inside the NUC.

wuxyintel replied:
Ok, list "Intel NUC Kit NUC7i7DNHE" here, thanks!

leilu replied:
Fix in next version

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,247 @@
+   Ubuntu installer creates 3 disk partitions on the on-board NVMe SSD. By
+   default, the GRUB bootloader is installed on the EFI System Partition (ESP),
+   used to bootstrap the ACRN hypervisor.
+#. After installing the Ubuntu OS, power off the NUC7i7DNB board, attach the
```
```suggestion
#. After installing the Ubuntu OS, power off the NUC, attach the
```

leilu replied:
Fix in next version

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,247 @@
+#. After installing the Ubuntu OS, power off the NUC7i7DNB board, attach the
+   SATA disk and storage device with USB interface to the board. Power on the
+   board and make sure it boots the Ubuntu OS from the NVMe SSD. Plug in the
+   removable disk with kernel image into the NUC7i7DNB, then copy the loadable
```
```suggestion
   removable disk with kernel image into the NUC, then copy the loadable
```

leilu replied:
Fix in next version

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,247 @@
+
+#. Before building the ACRN hypervisor, you need to figure out the I/O
+   address of the serial port, and the PCI BDF addresses of the SATA controller
+   and the USB controllers on NUC7i7DNB board. Enter the following command to
```
```suggestion
   and the USB controllers on the NUC. Enter the following command to
```

leilu replied:
Fix in next version

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,247 @@
+#. Before building the ACRN hypervisor, you need to figure out the I/O
+   address of the serial port, and the PCI BDF addresses of the SATA controller
+   and the USB controllers on NUC7i7DNB board. Enter the following command to
+   get the I/O addresses of the serial port. NUC7i7DNB boards support one serial
```
```suggestion
   get the I/O addresses of the serial port. The NUC supports one serial
```

leilu replied:
Fix in next version

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,247 @@
+
+   #. Copy ``acrn.32.out`` to a removable disk.
+
+   #. Plug in the removable disk into NUC7i7DNB.
```
```suggestion
   #. Plug in the removable disk into the NUC's USB port.
```

leilu replied:
Fix in next version

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,247 @@
+
+      $ sudo update-grub
+
+#. Reboot the NUC7i7DNB board. Select the "ACRN hypervisor Logical Partition
```
```suggestion
#. Reboot the NUC. Select the "ACRN hypervisor Logical Partition
```

leilu replied:
Fix in next version

dbkinder commented on doc/tutorials/setup_logical_partition_scenario_on_nuc7i7dnb.rst:
```diff
@@ -0,0 +1,247 @@
+
+#. Reboot the NUC7i7DNB board. Select the "ACRN hypervisor Logical Partition
+   Scenario" entry to boot the logical partition of the ACRN hypervisor on the
+   NUC7i7DNB display. The GRUB loader will boot the hypervisor, and the
```
```suggestion
   NUC's display. The GRUB loader will boot the hypervisor, and the
```

leilu replied:
Fix in next version

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_nuc.rst:
```diff
@@ -0,0 +1,247 @@
+
+              echo 'Loading hypervisor logical partition scenario ...'
+              multiboot --quirk-modules-after-kernel /boot/acrn.32.out
+              module /boot/bzImage
```
Is it correct that the kernel cmdline is defined here: https://github.com/projectacrn/acrn-hypervisor/blob/master/hypervisor/scenarios/logical_partition/vm_configurations.h#L31-L33? I would suggest we add a note to mention this as well in case someone wants to change it.

As a side note, the `VMx_CONFIG_OS_BOOTARG_*` macros are the only ones to use `bootarg` in its singular form. The rest of the code uses bootarg**s** (plural). This is not critical but I think it would be better to redefine these as `VM0_CONFIG_OS_BOOTARGS_*`.

wuxyintel replied:
Yes, it is fine to add this note.

gvancuts commented on doc/tutorials/setup_logical_partition_scenario_on_nuc.rst:
```diff
@@ -0,0 +1,247 @@
+
+   We recommend keeping the default values of items not mentioned above.
+
+#. Check GRUB boot loader name. 
```
I don't understand this step, what is its purpose?

wuxyintel replied:
@gvancuts  Thanks! On the current design, logical partition depends on the default boot loader (GRUB) and its name of ubuntu 16.04. Ubuntu installing maybe guarantee this, otherwise, the hypervisor will fails to boot. Only double check it here, do you think we can skip this double check?

leilu replied:
Update double check boot loader name steps.

leilu commented:
> We're really close, just a couple of small comments.
> 
> @dbkinder , I think it would be good to make this one appear next to the tutorial we have on setting up logical partition on UP2. I guess that would mean manually editing the `index.rst` file, or is there a different mechanism?
> 
> @leilu , will you be the one taking care of updating the same tutorial for UP2. It is outdated, see #3126
> 
> Thanks!
To make this one appear next to the tutorial we have on setting up logical partition on UP2, I think we could change the rst file name to "using_partiton_mode_on_nuc.rst".
To update the tutorial for UP2, may need @NanlinXie to assign the task.

gvancuts commented:
> @gvancuts Thanks! On the current design, logical partition depends on the default boot loader (GRUB) and its name of ubuntu 16.04. Ubuntu installing maybe guarantee this, otherwise, the hypervisor will fails to boot. Only double check it here, do you think we can skip this double check?

I think it's good to double-check this here. But it's not clear to me exactly what to check. You seem to be running a couple of commands and are talking about prefix. But I don't know which output to check exactly and what prefix to look for. If this can be changed in case a different multiboot boot loader is used, then it would also be good to add a note indicating where in the code this ought to be changed.

Thanks,
Geoffroy

gvancuts commented on doc/tutorials/using_partition_mode_on_nuc.rst:
```diff
@@ -0,0 +1,249 @@
+      In the head file
+      ``hypervisor/scenarios/logical_partition/vm_configurations.h``, the
+      ``VMx_CONFIG_OS_BOOTARG_*`` macros are the only ones to use bootarg in its
+      singular form. The rest of the code uses bootargs (plural).
```
This was meant as a comment about the code, i.e. it would be better that we change the name of the macros in our code. I wouldn't include this in the tutorial, just indicate which macros people need to edit if they want to change the kernel command-line parameters that are used by default.

leilu replied:
fixed at next version

gvancuts commented on doc/tutorials/using_partition_mode_on_nuc.rst:
```diff
@@ -0,0 +1,249 @@
+
+   .. note:: 
+
+      The VMs kernel cmdline is defined in head file
```
```suggestion
      The VMs kernel cmdline is defined in header file
```

leilu replied:
fixed at next version
